### PR TITLE
List ALL bpfPrograms

### PR DESCRIPTION
**Pull request by astoycos**
Works to finish #320 

Requires the two AYA commits from https://github.com/aya-rs/aya/pull/637

Existing `bpfctl list` functionality stays the same 

```bash
[astoycos@localhost bpfd]$ ./target/debug/bpfctl list
 Name  UUID                                  Type  Location                                                                         Metadata                                                                         
 pass  2d109bb0-5b8a-4ba4-b1f5-d5a852b166cb  xdp   image: { url: quay.io/bpfd-bytecode/xdp_pass:latest, pullpolicy: IfNotPresent }  { priority: 100, iface: eno1, position: 0, proceed_on: pass, dispatcher_return } 
```

To see ALL program info for ALL programs run `bpfctl list-all`

```
[astoycos@localhost bpfd]$ ./target/debug/bpfctl list-all
 Name             Id     Type  Loaded-At                 Tag               Gpl-compatible  Map-ids  Btf-id  Bytes_xlated  Jited  Bytes_jited  Bytes_memlock  Verified_insns 
 dump_bpf_map     693    26    2023-03-02T13:15:06-0500  749172daffada61f  true            [45]     154     264           true   287          4096           34             
 dump_bpf_prog    694    26    2023-03-02T13:15:06-0500  bc36dd738319ea32  true            [45]     154     528           true   715          4096           84             
 restrict_filesy  63130  29    2023-04-18T18:26:48-0400  aae89fa01fe7ee91  true            [62095]  69004   560           true   297          4096           67             
 sd_fw_egress     63149  8     2023-04-18T18:30:39-0400  6deef7357e7b4530  true            []       0       64            true   55           4096           8              
 sd_fw_ingress    63150  8     2023-04-18T18:30:39-0400  6deef7357e7b4530  true            []       0       64            true   55           4096           8              
 sd_fw_egress     63171  8     2023-04-18T18:31:51-0400  6deef7357e7b4530  true            []       0       64            true   55           4096           8              
 sd_fw_ingress    63172  8     2023-04-18T18:31:51-0400  6deef7357e7b4530  true            []       0       64            true   55           4096           8              
 sd_fw_egress     63173  8     2023-04-18T18:31:51-0400  6deef7357e7b4530  true            []       0       64            true   55           4096           8              
 sd_fw_ingress    63174  8     2023-04-18T18:31:51-0400  6deef7357e7b4530  true            []       0       64            true   55           4096           8              
 sd_fw_egress     63175  8     2023-04-18T18:31:51-0400  6deef7357e7b4530  true            []       0       64            true   55           4096           8              
 sd_fw_ingress    63176  8     2023-04-18T18:31:51-0400  6deef7357e7b4530  true            []       0       64            true   55           4096           8              
 sd_fw_egress     63177  8     2023-04-18T18:31:51-0400  6deef7357e7b4530  true            []       0       64            true   55           4096           8              
 sd_fw_ingress    63178  8     2023-04-18T18:31:51-0400  6deef7357e7b4530  true            []       0       64            true   55           4096           8              
                  73957  8     2023-06-28T12:13:39-0400  6deef7357e7b4530  true            []       0       64            true   55           4096           8              
                  73958  8     2023-06-28T12:13:39-0400  6deef7357e7b4530  true            []       0       64            true   55           4096           8              
 xdp              74409  6     2023-07-05T12:50:38-0400  90f686eb86991928  true            [68098]  84502   672           true   531          4096           96             
 pass             74410  28    2023-07-05T12:50:38-0400  4b9d1b2c14e87ce   true            [68099]  84503   96            true   67           4096           9      
```

This can be compared to bpftool output for correctness

```bash
[astoycos@localhost bpfd]$ sudo bpftool prog
693: tracing  name dump_bpf_map  tag 749172daffada61f  gpl
        loaded_at 2023-03-02T13:15:06-0500  uid 0
        xlated 264B  jited 287B  memlock 4096B  map_ids 45
        btf_id 154
        pids bpfd(46982)
694: tracing  name dump_bpf_prog  tag bc36dd738319ea32  gpl
        loaded_at 2023-03-02T13:15:06-0500  uid 0
        xlated 528B  jited 715B  memlock 4096B  map_ids 45
        btf_id 154
        pids bpfd(46982)
63130: lsm  name restrict_filesystems  tag aae89fa01fe7ee91  gpl
        loaded_at 2023-04-18T18:26:48-0400  uid 0
        xlated 560B  jited 297B  memlock 4096B  map_ids 62095
        btf_id 69004
        pids systemd(1), bpfd(46982)
63149: cgroup_skb  name sd_fw_egress  tag 6deef7357e7b4530  gpl
        loaded_at 2023-04-18T18:30:39-0400  uid 0
        xlated 64B  jited 55B  memlock 4096B
        pids systemd(1), bpfd(46982)
63150: cgroup_skb  name sd_fw_ingress  tag 6deef7357e7b4530  gpl
        loaded_at 2023-04-18T18:30:39-0400  uid 0
        xlated 64B  jited 55B  memlock 4096B
        pids systemd(1), bpfd(46982)
63171: cgroup_skb  name sd_fw_egress  tag 6deef7357e7b4530  gpl
        loaded_at 2023-04-18T18:31:51-0400  uid 0
        xlated 64B  jited 55B  memlock 4096B
        pids systemd(1), bpfd(46982)
63172: cgroup_skb  name sd_fw_ingress  tag 6deef7357e7b4530  gpl
        loaded_at 2023-04-18T18:31:51-0400  uid 0
        xlated 64B  jited 55B  memlock 4096B
        pids systemd(1), bpfd(46982)
63173: cgroup_skb  name sd_fw_egress  tag 6deef7357e7b4530  gpl
        loaded_at 2023-04-18T18:31:51-0400  uid 0
        xlated 64B  jited 55B  memlock 4096B
        pids systemd(1), bpfd(46982)
63174: cgroup_skb  name sd_fw_ingress  tag 6deef7357e7b4530  gpl
        loaded_at 2023-04-18T18:31:51-0400  uid 0
        xlated 64B  jited 55B  memlock 4096B
        pids systemd(1), bpfd(46982)
63175: cgroup_skb  name sd_fw_egress  tag 6deef7357e7b4530  gpl
        loaded_at 2023-04-18T18:31:51-0400  uid 0
        xlated 64B  jited 55B  memlock 4096B
        pids systemd(1), bpfd(46982)
63176: cgroup_skb  name sd_fw_ingress  tag 6deef7357e7b4530  gpl
        loaded_at 2023-04-18T18:31:51-0400  uid 0
        xlated 64B  jited 55B  memlock 4096B
        pids systemd(1), bpfd(46982)
63177: cgroup_skb  name sd_fw_egress  tag 6deef7357e7b4530  gpl
        loaded_at 2023-04-18T18:31:51-0400  uid 0
        xlated 64B  jited 55B  memlock 4096B
        pids systemd(1), bpfd(46982)
63178: cgroup_skb  name sd_fw_ingress  tag 6deef7357e7b4530  gpl
        loaded_at 2023-04-18T18:31:51-0400  uid 0
        xlated 64B  jited 55B  memlock 4096B
        pids systemd(1), bpfd(46982)
73957: cgroup_skb  tag 6deef7357e7b4530  gpl
        loaded_at 2023-06-28T12:13:39-0400  uid 0
        xlated 64B  jited 55B  memlock 4096B
        pids bpfd(46982), systemd(3466644)
73958: cgroup_skb  tag 6deef7357e7b4530  gpl
        loaded_at 2023-06-28T12:13:39-0400  uid 0
        xlated 64B  jited 55B  memlock 4096B
        pids bpfd(46982), systemd(3466644)
74409: xdp  name xdp  tag 90f686eb86991928  gpl
        loaded_at 2023-07-05T12:50:38-0400  uid 0
        xlated 672B  jited 531B  memlock 4096B  map_ids 68098
        btf_id 84502
        pids bpfd(46982)
74410: ext  name pass  tag 4b9d1b2c140e87ce  gpl
        loaded_at 2023-07-05T12:50:38-0400  uid 0
        xlated 96B  jited 67B  memlock 4096B  map_ids 68099
        btf_id 84503
        pids bpfd(46982)
```

### CODE REVIEW

File: bpfd/src/bpf.rs [hunk 1]
```diff
@@ -555,66 +557,133 @@ impl BpfManager {
...
                         },
                     ),
-                },
+                }
             })
             .collect();
-        Ok(programs)
+
+        loaded_programs()
+            .map(|p| {
```
Comment (dave-tucker) [inline bpfd/src/bpf.rs:651]: I would collect a `KernelProgramInfo` after we call `Load` - it can be stored in `Program` since it *shouldn't* change. One day we might want to add a reconciler that checks kernel state vs. desired state as if someone else unloads a program after we've loaded it we're in trouble anyway.

Back in relation to this code, it's then as simple as doing `loaded_programs().filter()` where the predicate function de-duplicates loaded programs by their ID. This is less error-prone than matching on name and/or tag.

File: bpfd/src/command.rs [hunk 2]
```diff
@@ -135,6 +135,14 @@ impl std::fmt::Display for Direction {
     }
 }
 
+#[derive(Debug, Clone)]
+pub(crate) struct ListInfo {
```
Comment (dave-tucker) [inline bpfd/src/command.rs:139]: The more I think on it, the more I don't think this is needed. Just create a ProgramInfo, and nest KernelProgramInfo inside of it.

File: bpfd/src/command.rs [hunk 3]
```diff
@@ -135,6 +135,14 @@ impl std::fmt::Display for Direction {
...
+    pub(crate) bpfd_info: Option<ProgramInfo>,
+    pub(crate) info: LinuxProgramInfo,
+}
+
+/// ProgramInfo stores information about bpf programs loaded on a system
+/// which are managed via bpfd.
 #[derive(Debug, Clone)]
 pub(crate) struct ProgramInfo {
     pub(crate) id: Uuid,
```
Comment (dave-tucker) [inline bpfd/src/command.rs:148]: You're going to ask - What UUID should I use here?
If we write one out to disk somewhere - i.e in `/var/run/bpfd`, we can use that as a namespace to generate v5 UUIDs from the ID and Tag of the programs in kernel (which are the uniquely identifying characteristics). You could even include load time in the hash if you wanted. 

```rust
use uuid::{Uuid}; // 1.4.0

fn main() {
    let host_uuid = Uuid::new_v4(); // populated at boot, stored on disk
    println!("Host UUID: {}", host_uuid);
    
    let id = 42u32;
    let tag = u64::MAX;
    
    let data = format!("{}:{:x}", id, tag); // could just use bytes instead of formatting
    println!("Data to be hashed: {}", data);
    let uuid = Uuid::new_v5(&host_uuid, &data.into_bytes()); 
    println!("{}", uuid);
}
```

This ID is stable across bpfd reboots because it would be re-generated using the same data.

**Review by dave-tucker — COMMENTED** [review]
Sorry for the patchy review - you've got raw stream of consciousness here while I digested the PR.

TL;DR we need to firm up the bpfd-side API a bit.
I think it's easy enough to generate `ProgramInfo` objects for all programs on the host, in response to a `List` from a client, then de-dupe that against a list of Programs that we loaded.
That would simplify the `List` command a lot IMHO.

I'm fine with this merging and doing *on demand* generation of a list of ProgramInfos, I think soon we'll want to spawn a thread that gathers kernel state + watches for changes. Not only is this more efficient, it will help us identify problems (i.e programs being unloaded etc...)

**Review by dave-tucker — CHANGES_REQUESTED** [review]
Sorry for the patchy review - you've got raw stream of consciousness here while I digested the PR.

TL;DR we need to firm up the bpfd-side API a bit.
I think it's easy enough to generate `ProgramInfo` objects for all programs on the host, in response to a `List` from a client, then de-dupe that against a list of Programs that we loaded.
That would simplify the `List` command a lot IMHO.

I'm fine with this merging and doing *on demand* generation of a list of ProgramInfos, I think soon we'll want to spawn a thread that gathers kernel state + watches for changes. Not only is this more efficient, it will help us identify problems (i.e programs being unloaded etc...)

File: proto/bpfd.proto [hunk 4]
```diff
@@ -15,13 +15,15 @@ message BytecodeImage {
...
+message NoLocation{}
+
 message LoadRequestCommon {
     oneof location {
         BytecodeImage image = 1;
         string file = 2;
     } ;
     string section_name = 3;
-    int32 program_type = 4;
+    uint32 program_type = 4;
```
[the remark this answers is not in the document]
Comment (astoycos) [reply proto/bpfd.proto:26]: This just mirrors the kernel type from https://elixir.bootlin.com/linux/latest/source/include/uapi/linux/bpf.h#L6301

 Keeping the types the same between our `Load` and `List` APIs makes keeps it simple to reconcile in our k8s setup, otherwise type munging is required

File: bpfd/src/bpf.rs [hunk 5]
```diff
@@ -555,66 +557,133 @@ impl BpfManager {
...
-                            tracepoint: p.info.tracepoint.to_string(),
+            .map(|(id, p)| {
+                // The key length is abbreviated to the first 16 bytes of the
+                // program name since that's what will be listed by the
+                // `aya::programs::loaded_programs()` API.
+                let prog_index = if p.section_name().len() > 16 {
+                    p.section_name()[0..16].to_string()
+                } else {
+                    p.section_name().to_string()
+                };
```
[the remark this answers is not in the document]
Comment (astoycos) [reply bpfd/src/bpf.rs:574]: This bit is going to need to change.... I was using prog "name" to match a program loaded by bpfd to a given `aya::ProgramInfo` but in reality that won't work, instead I'll need to use an actual unique id to do so.

Originally this was done though because the kernel limits the name to 16 bytes unfortunately see https://elixir.bootlin.com/linux/latest/source/include/uapi/linux/bpf.h#L1318

File: bpfd/src/bpf.rs (hunk 1 above)
Comment (astoycos) [reply bpfd/src/bpf.rs:651]: +1 this was the bit I couldn't figure out originally and at the time I was trying NOT to store a `KernelProgramInfo` in Program.

File: bpfd/src/bpf.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (astoycos) [reply bpfd/src/bpf.rs]: Not relevant anymore using kernel ID as the unique ident for match a bpfd prog to it's kernel meta

File: proto/bpfd.proto [hunk 6]
```diff
@@ -76,25 +78,38 @@ message UnloadRequest {
 message UnloadResponse {}
 
 message ListRequest {
-    optional int32 program_type = 1;
+    optional uint32 program_type = 1;
+    optional bool bpfd_programs_only = 2;
```
[the remark this answers is not in the document]
Comment (astoycos) [reply proto/bpfd.proto:82]: Done

File: bpfctl/src/main.rs [hunk 7]
```diff
@@ -204,26 +206,47 @@ struct GlobalArg {
     value: Vec<u8>,
 }
 
-struct ProgTable(Table);
+struct BpfdTable(Table);
```
[the remark this answers is not in the document]
Comment (astoycos) [reply bpfctl/src/main.rs]: Reverted back

File: bpfctl/src/main.rs [hunk 8]
```diff
@@ -204,26 +206,47 @@ struct GlobalArg {
...
 
         table.load_preset(comfy_table::presets::NOTHING);
-        table.set_header(vec!["UUID", "Type", "Name", "Location", "Metadata"]);
-        ProgTable(table)
+        table.set_header(vec!["Name", "UUID", "Type", "Location", "Metadata"]);
+
+        let mut meta_info = Table::new();
+
+        meta_info.load_preset(comfy_table::presets::NOTHING);
+        meta_info.set_header(vec![
```
[the remark this answers is not in the document]
Comment (astoycos) [reply bpfctl/src/main.rs]: Done

File: bpfd/src/command.rs (hunk 2 above)
Comment (astoycos) [reply bpfd/src/command.rs]: Done, now KernelProgramInfo is an integral part to our whole workflow

File: bpfd/src/multiprog/xdp.rs [hunk 9]
```diff
@@ -204,6 +204,9 @@ impl XdpDispatcher {
                 let target_fn = format!("prog{i}");
 
                 ext.load(dispatcher.fd().unwrap(), &target_fn)?;
+                debug!("XDP KERNEL INFO {:?}", ext.program_info());
```
Comment (astoycos) [inline bpfd/src/multiprog/xdp.rs:207]:
```suggestion
```

File: bpfd/src/rpc.rs [hunk 10]
```diff
@@ -230,61 +231,126 @@ impl Loader for BpfdLoader {
...
-                            crate::command::AttachInfo::Uprobe(info) => {
-                                AttachInfo::UprobeAttachInfo(UprobeAttachInfo {
-                                    fn_name: info.fn_name,
-                                    offset: info.offset,
-                                    target: info.target,
-                                    pid: info.pid,
-                                    namespace: info.namespace,
-                                })
-                            }
+                        debug!("RESULTS {:?}", r.clone());
```
Comment (astoycos) [inline bpfd/src/rpc.rs:234]:
```suggestion
```

File: bpfd/src/rpc.rs [hunk 11]
```diff
@@ -230,61 +231,126 @@ impl Loader for BpfdLoader {
...
+                        let loc;
+                        let attach_info;
+                        let name;
+                        // With bpfd "multi-attach" programs are loaded as
+                        // extensions to a dispatcher. The kernel see's these
+                        // programs as type "BPF_PROG_TYPE_EXT" rather than
+                        // their actual type.  Therefore we check here if the
+                        // program is owned by bpfd and use our stored "real"
+                        // type for the filtering. Otherwise we use what's stored
```
Comment (astoycos) [inline bpfd/src/rpc.rs:243]:
```suggestion
                        // type.
```

File: bpfctl/src/main.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (astoycos) [reply bpfctl/src/main.rs]: Well mostly done 

I'd like to do 

> Provide JSON as an output option in all cases so you can use jq to render what you want.

In a follow up if that's alright (I'll make an issue)

File: bpfctl/src/main.rs [hunk 12]
```diff
@@ -640,7 +719,26 @@ async fn execute_request(command: &Commands, channel: Channel) -> anyhow::Result
...
+
+            for r in response.results {
+                if let Err(e) = table.add_response_prog(r) {
+                    bail!(e)
+                }
+            }
+            println!("{table}");
+        }
+
+        Commands::ListAll {} => {
```
[the remark this answers is not in the document]
Comment (astoycos) [reply bpfctl/src/main.rs]: Done

File: bpfd/src/command.rs [hunk 13]
```diff
@@ -208,6 +235,26 @@ pub(crate) enum Program {
     Uprobe(UprobeProgram),
 }
 
+impl Program {
+    pub(crate) fn section_name(&self) -> &str {
```
[the remark this answers is not in the document]
Comment (astoycos) [reply bpfd/src/command.rs]: Removed this method don't need it :)

**Comment by astoycos** [comment]
@dave-tucker  The new API looks like this 

1. List bpfd programs
```bash
[astoycos@localhost bpfd]$ ./target/debug/bpfctl list
 Kernel_ID  Bpfd_UUID                             Name      Type    Load_Time                
 75299      dd701cb0-0ae9-4caa-a78d-0087b102f6cd  pass      xdp     2023-07-10T14:41:42-0400 
 75300      453f7f9c-c513-467e-82a0-961e086fda87  retprobe  kprobe  2023-07-10T14:41:46-0400 
 75302      776778ad-ed19-4940-a844-78dd00d201ae  pass      tc      2023-07-10T14:41:56-0400 
```
2. List all programs
```bash
[astoycos@localhost bpfd]$ ./target/debug/bpfctl list --all
 Kernel_ID  Bpfd_UUID                             Name             Type        Load_Time                
 693                                              dump_bpf_map     tracing     2023-03-02T13:15:06-0500 
 694                                              dump_bpf_prog    tracing     2023-03-02T13:15:06-0500 
 63130                                            restrict_filesy  lsm         2023-04-18T18:26:48-0400 
 63149                                            sd_fw_egress     cgroup_skb  2023-04-18T18:30:39-0400 
 63150                                            sd_fw_ingress    cgroup_skb  2023-04-18T18:30:39-0400 
 63171                                            sd_fw_egress     cgroup_skb  2023-04-18T18:31:51-0400 
 63172                                            sd_fw_ingress    cgroup_skb  2023-04-18T18:31:51-0400 
 63173                                            sd_fw_egress     cgroup_skb  2023-04-18T18:31:51-0400 
 63174                                            sd_fw_ingress    cgroup_skb  2023-04-18T18:31:51-0400 
 63175                                            sd_fw_egress     cgroup_skb  2023-04-18T18:31:51-0400 
 63176                                            sd_fw_ingress    cgroup_skb  2023-04-18T18:31:51-0400 
 63177                                            sd_fw_egress     cgroup_skb  2023-04-18T18:31:51-0400 
 63178                                            sd_fw_ingress    cgroup_skb  2023-04-18T18:31:51-0400 
 75294                                            dispatcher       tc          2023-07-10T14:34:21-0400 
 75298                                            xdp              xdp         2023-07-10T14:41:41-0400 
 75299      dd701cb0-0ae9-4caa-a78d-0087b102f6cd  pass             xdp         2023-07-10T14:41:42-0400 
 75300      453f7f9c-c513-467e-82a0-961e086fda87  retprobe         kprobe      2023-07-10T14:41:46-0400 
 75301                                            dispatcher       tc          2023-07-10T14:41:55-0400 
 75302      776778ad-ed19-4940-a844-78dd00d201ae  pass             tc          2023-07-10T14:41:56-0400 
```
3. GET a bpfd program
```bash
[astoycos@localhost bpfd]$ ./target/debug/bpfctl get 75302

#################### Bpfd State ####################

UUID:                               776778ad-ed19-4940-a844-78dd00d201ae                              
Image:                              quay.io/bpfd-bytecode/tc_pass:latest 
ImagePullpolicy:                    IfNotPresent             
Priority:                           100
Iface:                              eno1
Position:                           0
Direction:                          in
Proceed_on:                         pipe, dispatcher_return

#################### Kernel State ##################

Kernel_ID:                          75302
Name:                               pass
Type:                               tc
Loaded At:                          2023-07-10T14:41:56-0400
Tag:                                9b2c38d37350bfff
GPL Compatible:                     true
Map IDs:                            [68907]
BTF ID:                             86020
Size Translated (bytes):            96
JITed:                              true
Size JITed (bytes):                 64
Kernel Allocated Memory (bytes):    4096
Verified Instruction Count:         9
```
4. Get another Program
```
[astoycos@localhost bpfd]$ ./target/debug/bpfctl get 693

#################### Bpfd State ####################
NONE

#################### Kernel State ##################

Kernel_ID:                          693
Name:                               dump_bpf_map
Type:                               tracing
Loaded At:                          2023-03-02T13:15:06-0500
Tag:                                749172daffada61f
GPL Compatible:                     true
Map IDs:                            [45]
BTF ID:                             154
Size Translated (bytes):            264
JITed:                              true
Size JITed (bytes):                 287
Kernel Allocated Memory (bytes):    4096
Verified Instruction Count:         34
```

This PR is still pointing to https://github.com/astoycos/aya Where we are carrying 

https://github.com/aya-rs/aya/pull/520
https://github.com/aya-rs/aya/pull/560
https://github.com/aya-rs/aya/pull/637

Which have all now been rebased (and fixed up) to aya:main 😩  Once you LGTM my latest aya changes (#637) I'll push the rebased aya branch to bpfd-dev/aya.

I also have some follow up issues to this I need to make
1. json output for all out bpfctl listing commands
2. Add K8s support for this new API
3. Rename all references to section_name(I already started this but decided it would make this PR huge)
4.  <Maybe some others I'm forgetting right now>

File: bpfd/src/bpf.rs [hunk 14]
```diff
@@ -243,15 +242,25 @@ impl BpfManager {
                 Err(e)
             })?;
         self.dispatchers.insert(did, dispatcher);
+
+        // update programs with now populated kernel info
```
Comment (astoycos) [inline bpfd/src/bpf.rs:246]: @dave-tucker I'm not sure how to do this any better ATM without dramatically changing some of our dataflow/function design.... Maybe I just can't see if because I'm tired could you provide some input?

File: bpfd/src/command.rs [hunk 15]
```diff
@@ -144,6 +152,25 @@ pub(crate) struct ProgramInfo {
...
 }
 
+/// LinuxProgramInfo stores information about ALL bpf programs loaded
+/// on a system.
+#[derive(Debug, Clone)]
+pub(crate) struct LinuxProgramInfo {
+    pub(crate) id: u32,
+    pub(crate) name: String,
+    pub(crate) program_type: u32,
+    pub(crate) loaded_at: String,
```
[the remark this answers is not in the document]
Comment (astoycos) [reply bpfd/src/command.rs:147]: Done, Now we're just doing string formatting here :)

File: bpfd/src/command.rs (hunk 3 above)
Comment (astoycos) [reply bpfd/src/command.rs]: I never needed to generate a UUID, since we can't really do that deterministically since now users can provide their own UUID.  Therefore, I kept the bpfd UUID and Kernel_ID as  two separate entities.

File: bpfd/src/bpf.rs (hunk 1 above)
Comment (astoycos) [reply bpfd/src/bpf.rs:624]: Additionally you had asked me to use `filter` here on `loaded_programs`  with the kernel_IDs of the programs being handled by kernel ids however using a map lookup to track the bpfd programs seemed more performant since it required only iterating through the bpfd Programs and loaded programs once 

While using filter to de-duplicate based on a vector of map IDs would have required us to iterate through both lists multiple times. 

Again please point it out if I'm missing something here as I very well may be.

File: bpfctl/src/main.rs [hunk 16]
```diff
@@ -43,7 +43,21 @@ enum Commands {
...
+    Get { kernel_id: u32 },
+}
+
+#[derive(Args)]
+struct ListArgs {
+    /// Optional: Program type to list
+    /// Example: --program-type xdp
+    #[clap(short, long, verbatim_doc_comment)]
+    program_type: Option<ProgramType>,
+
```
Comment (Billy99) [inline bpfctl/src/main.rs:57]:
```suggestion
    /// Optional: Program type to list
    /// Example: --program-type xdp
    /// [possible values: unspec, socket-filter, kprobe, tc, sched-act,
    ///                   tracepoint, xdp, perf-event, cgroup-skb,
    ///                   cgroup-sock, lwt-in, lwt-out, lwt-xmit, sock-ops,
    ///                   sk-skb, cgroup-device, sk-msg, raw-tracepoint,
    ///                   cgroup-sock-addr, lwt-seg6-local, lirc-mode2,
    ///                   sk-reuseport, flow-dissector, cgroup-sysctl,
    ///                   raw-tracepoint-writable, cgroup-sockopt, tracing,
    ///                   struct-ops, ext, lsm, sk-lookup, syscall]
    #[clap(short, long, verbatim_doc_comment, hide_possible_values=true)]
    program_type: Option<ProgramType>,

```

Not a fan of the original helptext output (wraps on my terminal):
```
$ bpfctl list --help
List all BPF programs loaded via bpfd

Usage: bpfctl list [OPTIONS]

Options:
  -p, --program-type <PROGRAM_TYPE>  Optional: Program type to list
                                     Example: --program-type xdp [possible values: unspec, socket-filter, kprobe, tc, sched-a
ct, tracepoint, xdp, perf-event, cgroup-skb, cgroup-sock, lwt-in, lwt-out, lwt-xmit, sock-ops, sk-skb, cgroup-device, sk-msg,
 raw-tracepoint, cgroup-sock-addr, lwt-seg6-local, lirc-mode2, sk-reuseport, flow-dissector, cgroup-sysctl, raw-tracepoint-wr
itable, cgroup-sockopt, tracing, struct-ops, ext, lsm, sk-lookup, syscall]
  -a, --all                          
  -h, --help                         Print help
```
With this change, it now looks like:
```
$ bpfctl list --help
List all BPF programs loaded via bpfd

Usage: bpfctl list [OPTIONS]

Options:
  -p, --program-type <PROGRAM_TYPE>  Optional: Program type to list
                                     Example: --program-type xdp
                                     [possible values: unspec, socket-filter, kprobe, tc, sched-act,
                                                       tracepoint, xdp, perf-event, cgroup-skb,
                                                       cgroup-sock, lwt-in, lwt-out, lwt-xmit, sock-ops,
                                                       sk-skb, cgroup-device, sk-msg, raw-tracepoint,
                                                       cgroup-sock-addr, lwt-seg6-local, lirc-mode2,
                                                       sk-reuseport, flow-dissector, cgroup-sysctl,
                                                       raw-tracepoint-writable, cgroup-sockopt, tracing,
                                                       struct-ops, ext, lsm, sk-lookup, syscall]
  -a, --all                          
  -h, --help                         Print help
```

I know there is more maintenance with this approach, but it makes it so much easier to read. And I don't think the kernel ProgramType will change that often, and if it does and the helptext part is missed, it will still filter properly on the new value, just the list will be outdated.

File: bpfctl/src/main.rs [hunk 17]
```diff
@@ -43,7 +43,21 @@ enum Commands {
     /// Unload a BPF program using the UUID.
     Unload(UnloadArgs),
     /// List all BPF programs loaded via bpfd.
-    List,
+    List(ListArgs),
+    /// Get a program's metadata information specified by kernel id.
+    Get { kernel_id: u32 },
```
Comment (Billy99) [inline bpfctl/src/main.rs:48]: Add a new issue/follow-up item for both `bpfctl get` and `bpfctl unload` to take either a KernelId or UUID. One parameter, if its a number we assume it to be a KernelId, o.w. assume it's a UUID/

File: bpfd-api/src/bpfd.v1.rs [hunk 18]
```diff
@@ -142,15 +147,37 @@ pub mod list_response {
...
     #[derive(Clone, PartialEq, ::prost::Message)]
     pub struct ListResult {
-        #[prost(string, tag = "1")]
-        pub id: ::prost::alloc::string::String,
-        #[prost(string, optional, tag = "2")]
-        pub section_name: ::core::option::Option<::prost::alloc::string::String>,
-        #[prost(int32, tag = "5")]
-        pub program_type: i32,
-        #[prost(oneof = "list_result::Location", tags = "3, 4")]
+        #[prost(string, optional, tag = "1")]
```
[the remark this answers is not in the document]
Comment (astoycos) [reply bpfd-api/src/bpfd.v1.rs:150]: So this is the rust bindings that are generated from the [proto file ](https://github.com/bpfd-dev/bpfd/blob/main/proto/bpfd.proto) I would imagine that those tags are automatically generated by tonic on the rust struct.

File: bpfctl/src/main.rs [hunk 19]
```diff
@@ -9,7 +9,7 @@ use bpfd_api::{
     config::{self, Config},
     util::directories::*,
     v1::{
-        list_response,
+        list_response::{self, list_result::Location},
```
[the remark this answers is not in the document]
Comment (astoycos) [reply bpfctl/src/main.rs:12]: Done

File: bpfctl/src/main.rs (hunk 17 above)
Comment (astoycos) [reply bpfctl/src/main.rs:48]: Good point, done!

File: bpfctl/src/main.rs (hunk 16 above)
Comment (astoycos) [reply bpfctl/src/main.rs:57]: Sounds good to me, I don't think there's much more maintenance I doubt many bpf program types will be added in the future.

File: bpfctl/src/main.rs [hunk 20]
```diff
@@ -211,24 +232,58 @@ impl ProgTable {
...
-        match prog_type {
+        let location = match &r.clone().location.unwrap() {
+            // Cast imagePullPolicy into it's concrete type so we can easily print.
+            Location::Image(i) => format!(
+                r#"Image:                              {} 
+ImagePullpolicy:                    {}"#,
+                i.url,
+                TryInto::<ImagePullPolicy>::try_into(i.image_pull_policy)?
+            ),
+            Location::File(p) => format!(r#"Path:                           {p}"#),
```
Comment (Billy99) [inline bpfctl/src/main.rs:282]:
```suggestion
            Location::File(p) => format!(r#"Path:                               {p}"#),
```
Looks like for a file, you need a few more spaces:
```
$ bpfctl get 5862

#################### Bpfd State ####################

UUID:                               6ea3e8b2-5bfe-45f3-8968-da1aefd9f106                              
Path:                           /run/bpfd/examples/go-xdp-counter/bpf_bpfel.o             
Priority:                           100
Iface:                              vethff657c7
:
```

File: bpfctl/src/main.rs [hunk 21]
```diff
@@ -43,7 +43,28 @@ enum Commands {
     /// Unload a BPF program using the UUID.
     Unload(UnloadArgs),
     /// List all BPF programs loaded via bpfd.
-    List,
+    List(ListArgs),
+    /// Get a program's metadata information specified by kernel id.
+    Get { kernel_id: u32 },
```
Comment (Billy99) [inline bpfctl/src/main.rs:48]: nit: Since the only parameter is defined here and not in a struct below, there is no helptext for `kernel_id`:

```
$ bpfctl get --help
Get a program's metadata information specified by kernel id

Usage: bpfctl get <KERNEL_ID>

Arguments:
  <KERNEL_ID>  

Options:
  -h, --help  Print help
```

**Review by Billy99 — COMMENTED** [review]
Few more nits

File: bpfctl/src/main.rs (hunk 20 above)
Comment (astoycos) [reply bpfctl/src/main.rs:282]: Ah great catch

File: bpfctl/src/main.rs (hunk 21 above)
Comment (astoycos) [reply bpfctl/src/main.rs:48]: fixed to 

```
[astoycos@localhost bpfd]$ ./target/debug/bpfctl get --help
Get a program's metadata information specified by kernel id

Usage: bpfctl get <KERNEL_ID>

Arguments:
  <KERNEL_ID>  A BPF program's kernel id

Options:
  -h, --help  Print help
```

Is that alright?

Comment (Billy99) [reply bpfctl/src/main.rs:48]: Yes

File: proto/bpfd.proto [hunk 22]
```diff
@@ -76,25 +78,38 @@ message UnloadRequest {
...
+    bool gpl_compatible = 15;
+    repeated uint32 map_ids = 16;
+    uint32 btf_id = 17;
+    uint32 bytes_xlated = 18;
+    bool jited = 19;
+    uint32 bytes_jited = 20;
+    uint32 bytes_memlock = 21;
+    uint32 verified_insns = 22;
   } 
   repeated ListResult results = 10;
```
Comment (Billy99) [inline proto/bpfd.proto:114]: Should this be `23`?

File: bpfd-api/src/bpfd.v1.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (astoycos) [reply bpfd-api/src/bpfd.v1.rs:150]: Fixed

File: proto/bpfd.proto (hunk 22 above)
Comment (astoycos) [reply proto/bpfd.proto]: For consistency yes 👍

**Comment by astoycos** [comment]
The current CI failure is transient/ a problem with ubuntu package mirrors.. will take a look tomorrow.

**Comment by astoycos** [comment]
Removing hold, I went ahead and updated `github.com/bpfd-dev/aya:main` since we're only carrying https://github.com/aya-rs/aya/pull/637 now

**Review by Billy99 — APPROVED** [review]
LGTM

**Comment by astoycos** [comment]
Also opened this to refactor the `bpfctl get` output using comfy table https://github.com/bpfd-dev/bpfd/issues/601

File: bpfd/src/bpf.rs (hunk 14 above)
Comment (dave-tucker) [reply bpfd/src/bpf.rs:246]: You could get the info at load time here: https://github.com/astoycos/bpfd/blob/066e596a531d155a261be41fc0cf035fb39b2589/bpfd/src/multiprog/xdp.rs#L148

**Review by dave-tucker — APPROVED** [review]
